ntp-proto/src/packet/v5.rs

```diff
@@ -0,0 +1,409 @@
+use crate::packet::error::ParsingError;
+use crate::{NtpDuration, NtpLeapIndicator, NtpTimestamp};
+
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum NtpMode {
+    Request = 3,
+    Response = 4,
+}
+
+impl NtpMode {
+    fn from_bits(bits: u8) -> Result<Self, ParsingError<std::convert::Infallible>> {
+        Ok(match bits {
+            3 => Self::Request,
+            4 => Self::Response,
+            _ => return Err(ParsingError::MalformedMode),
+        })
+    }
+
+    fn to_bits(self) -> u8 {
+        match self {
+            Self::Request => 3,
+            Self::Response => 4,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn is_request(&self) -> bool {
+        self == &Self::Request
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn is_response(&self) -> bool {
+        self == &Self::Response
+    }
+}
+
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum NtpTimescale {
+    Utc = 0,
+    Tai = 1,
+    Ut1 = 2,
+    LeadSmearedUtc = 3,
+}
+
+impl NtpTimescale {
+    fn from_bits(bits: u8) -> Result<Self, ParsingError<std::convert::Infallible>> {
+        Ok(match bits {
+            0 => Self::Utc,
+            1 => Self::Tai,
+            2 => Self::Ut1,
+            3 => Self::LeadSmearedUtc,
+            _ => return Err(ParsingError::MalformedTimescale),
+        })
+    }
+
+    fn to_bits(self) -> u8 {
+        match self {
+            Self::Utc => 0,
+            Self::Tai => 1,
+            Self::Ut1 => 2,
+            Self::LeadSmearedUtc => 3,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct NtpEra(pub u8);
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct NtpFlags {
+    unknown_leap: bool,
+    interleaved_mode: bool,
+}
+
+impl NtpFlags {
+    fn from_bits(bits: [u8; 2]) -> Result<Self, ParsingError<std::convert::Infallible>> {
+        if bits[0] != 0x00 || bits[1] & 0xFC != 0 {
+            return Err(ParsingError::InvalidFlags);
+        }
+
+        Ok(Self {
+            unknown_leap: bits[1] & 0x01 != 0,
+            interleaved_mode: bits[1] & 0x02 != 0,
```

Comment:
random suggestion: to me bitwise operations with binary literals make more sense. Hex always needs some mental conversion

Reply:
I would agree, but the spec lists these specifically as hex. So I am not sure if we should go more for matching the spec or be easier to see, or maybe changing the spec to binary as well?

Reply:
yeah, apparently C has binary literals since C99. I'd appreciate that change (in the spec too)